src/BinSkim.Rules/PERules/BA2006.BuildWithSecureTools.cs

```diff
@@ -299,93 +241,124 @@
                         string.Join(", ", sorted)));
         }
 
-        private string BuildCompilerIdentifier(ObjectModuleDetails omDetails)
-        {
-            return omDetails.CompilerName + ":" + omDetails.Language + ":" + omDetails.CompilerBackEndVersion;
-        }
-
-        private string BuildMinimumCompilersList(BinaryAnalyzerContext context, Dictionary<Language, List<ObjectModuleDetails>> languageToBadModules)
+        internal void GenerateMessageParametersAndLog(BinaryAnalyzerContext context, Dictionary<Language, List<ObjectModuleDetails>> languageToBadModules)
         {
+            var sb = new StringBuilder();
             var languages = new List<string>();
-
-            foreach (Language language in languageToBadModules.Keys)
+            foreach (KeyValuePair<Language, List<ObjectModuleDetails>> kp in languageToBadModules.OrderBy(l => l.Key))
             {
-                Version version = context.Policy.GetProperty(MinimumToolVersions)[language.ToString()];
-                languages.Add($"{language} ({version})");
+                sb.Append(kp.Value.CreateOutputCoalescedByCompiler());
+
+                Version version = RetrieveMinimumCompilerVersionByLanguage(context, kp.Key);
```

Comment:
>kp.Key [](http://example.com/codeflow?start=84&length=6)

when you embed things like 'kp.Key' as an argument, you lose readability. For this code, I might have passed this using the parameter naming convention or created a local that retains information that the key is the language.